lib/ansible/modules/cloud/podman/podman_container.py

```diff
@@ -0,0 +1,1196 @@
+class PodmanManager:
+
+    def __init__(self, module, results):
+
+        super(PodmanManager, self).__init__()
+
+        self.module = module
+        self.results = results
+        self.name = self.module.params.get('name')
+        self.executable = self.module.get_bin_path(module.params.get('executable'), required=True)
+        self.image = self.module.params.get('image')
+        if self.image:
+            if ensure_image_exists(self.module, self.image):
+                self.results['actions'].append('pulled image %s' % self.image)
+        self.state = self.module.params.get('state')
+        self.restart = self.module.params.get('force_restart')
+        self.recreate = self.module.params.get('recreate')
+        self.container = PodmanContainer(self.module, self.name, self.module.params)
+
+    def update_container_result(self, changed=True):
+        facts = self.container.get_info()
+        self.results.update({'changed': changed, 'container': facts,
+                             'ansible_facts': {'podman_container': facts}})
+        return self.results
+
+    def execute(self):
+        if self.state in ['present', 'started']:
+            if not self.container.exists and not self.image:
+                self.module.fail_json(msg='Cannot create container when image is not specified!')
+            if self.container.running and (self.container.different or self.recreate):
+                if not self.image:
+                    return self.fail(msg='Cannot create container when image is not specified!')
+                self.container.recreate()
+                self.results['actions'].append('recreated %s' % self.container.name)
+                return self.update_container_result()
+            elif self.container.running and not self.container.different:
+                if self.restart:
+                    self.container.restart()
+                    self.results['actions'].append('restarted %s' % self.container.name)
+                    return self.update_container_result()
+                return self.results
+            elif self.container.stopped and self.container.different:
+                self.container.recreate()
+                self.results['actions'].append('recreated %s' % self.container.name)
+                return self.update_container_result()
+            elif self.container.stopped and not self.container.different:
+                self.container.start()
+                self.results['actions'].append('started %s' % self.container.name)
+                return self.update_container_result()
+            elif not self.container.exists:
+                self.container.run()
+                self.results['actions'].append('started %s' % self.container.name)
+                return self.update_container_result()
+
+        if self.state == 'stopped':
+            if not self.container.exists and not self.image:
+                return self.fail(msg='Cannot create container when image is not specified!')
```

Comment:
Why would you create a container when asked to stop?

Reply:
it means I need to create a container, but not to run it (in "stopped"/"created" state), follow here docker logic

Reply:
It just looks odd to me. I'd drop `and not self.image` part.

In general, it's good to have your code structured with separate layers. When the user input enters your app flow, in the beginning, you should do its initial validation and fail immediately if needed. As it gets bypassed deeper into the next layers you can be sure that the input is already sanitized and 100% valid so that you don't need to do extra checks in the app internals, focusing on the actual logic.

Reply:
Yeah, I even prepared a section below to work on all params limits. The problem is that it sometimes depends on current state and not just parameters logic. For example if state is stopped, but container doesn't exist - I need the image to be defined, if container exists - I don't need it be defined.
Everything that doesn't depend on current state I'll filter out in main, of course.

Reply:
Maybe you just need to separate such things into functions so that low-level implementation details wouldn't leak into the higher-level abstraction confusing the reader.

Reply:
TBH, except of this check I don't have anymore I think, not sure if it worth the effort.

Reply:
k